src/NPZ.jl

```diff
@@ -58,36 +57,37 @@ end
 # be fixed by rehashing the Dict when the module is
 # loaded.
 
-readle(ios::IO, ::Type{UInt16}) = htol(read(ios, UInt16))
+readle(ios::IO, ::Type{T}) where T = ltoh(read(ios, T)) #ltoh is inverse of htol
 
-function writele(ios::IO, x::Vector{UInt8})
-    n = write(ios, x)
-    if n != length(x)
-        error("short write")
-    end
-    n
+function writecheck(io::IO, x::Any)
+    n = write(io, x) # returns size in bytes
+    n == sizeof(x) || error("short write") # sizeof is size in bytes
 end
 
-writele(ios::IO, x::String) = writele(ios, convert(Vector{UInt8}, x))
-writele(ios::IO, x::UInt16) = writele(ios, reinterpret(UInt8, [htol(x)]))
+# Endianness only pertains to multi-byte things
+writele(ios::IO, x::AbstractVector{UInt8}) = writecheck(ios, x)
+# NPY headers only have ascii strings (bytes) so endianness doesn't matter
+writele(ios::IO, x::AbstractString) = writecheck(ios, codeunits(x))
 
-function parsechar(s::String, c::Char)
+writele(ios::IO, x::UInt16) = writecheck(ios, htol(x))
+
+function parsechar(s::AbstractString, c::Char)
```

Comment:
This function was already a bit wrong in that it assumes `c` is an ASCII character. If it's not, it'll crash:
```
julia> s = "\u2200 x \u2203 y"
"∀ x ∃ y"

julia> typeof(s)
String

julia> s[1]
'∀': Unicode U+2200 (category Sm: Symbol, math)

julia> s[2:end]
ERROR: UnicodeError: invalid character index 2 (0x88 is a continuation byte)
Stacktrace:
 [1] macro expansion at ./REPL.jl:97 [inlined]
 [2] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
```
Now that it's `AbstractString`, it may be UTF-16 and become always wrong.

Reply:
Hmm, it was already broken. `String` can be (and is) utf8, and you just demonstrated that it breaks on utf8.

To demonstrate...
```julia
julia> foo(s::String) = s
foo (generic function with 1 method)
julia> s = "♨★  𝕦ᑎ𝐢ς𝐎𝐝є!!  ☯✌"
"♨★  𝕦ᑎ𝐢ς𝐎𝐝є!!  ☯✌"
julia> bar(s::String) = s[2]
bar (generic function with 1 method)
julia> bar(s)
ERROR: StringIndexError("♨★  𝕦ᑎ𝐢ς𝐎𝐝є!!  ☯✌", 2)
Stacktrace:
 [1] string_index_err(::String, ::Int64) at ./strings/string.jl:12
 [2] getindex_continued(::String, ::Int64, ::UInt32) at ./strings/string.jl:216
 [3] getindex at ./strings/string.jl:209 [inlined]
 [4] bar(::String) at ./REPL[31]:1
 [5] top-level scope at none:0
```